XWIKI-21375: Videos on the help page don't have alternatives

Sereza7 opened the pull request:
## Jira
https://jira.xwiki.org/browse/XWIKI-21375
## Forum discussion
[Handle videos in the help page](https://forum.xwiki.org/t/handle-videos-in-the-help-page/13220)
## PR Changes
* Changed the list items to anchors sending to different documentation pages
* Added a hint to explicitly set the text alternative of the videos
## View
We can see that now all list items under the videos are anchors that redirect to various sections of the documentation. 
![21375-afterPR](https://github.com/xwiki/xwiki-platform/assets/28761965/4f62a2d9-fbc0-42e5-b5de-45744bcbbd99)
## Notes
The way I implemented the solution we converged to in the discussion is based upon [this example in a proposal of atomic rule for ACT](https://www.w3.org/WAI/standards-guidelines/act/rules/fd26cf/proposed/#test-cases). There is no designated role for alternative to videos in aria, we rely on the user's understanding.
I decided to add a non-visual hint to highlight the bond between the video and its text counterpart. This element would clutter the visual UI, and fills a role for AT users that presentation handles visually.

Here are all the pairs label + URL added in this PR:
*  [Discover the elements displayed on your homepage](https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/)
* [Start browsing your wiki](https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/Features/Navigate/) 
* [Learn how to change your wiki colour theme ](https://extensions.xwiki.org/xwiki/bin/view/Extension/Flamingo%20Theme%20Application#HHowtoselectatheme)
* [ Learn how to add panels and columns to your wiki look ](https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/ChangingTheLogoAndThePanels#HChangingyourwiki27spanels)
*  [Learn how to add widgets to a wiki ](https://extensions.xwiki.org/xwiki/bin/view/Extension/Dashboard%20Macro#HAddinggadgets)
*  [Discover how to create a page ](https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/CreatingAPage)
*  [Learn how to use the WYSIWYG editor ](https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/EditingAPage)
* [Discover the view and edit modes ](https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/Features/KeyboardShortcuts#HViewMode)
*  [Discover how to import Office Documents to your wiki ](https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/Features/Imports#HImportingOfficedocuments)
*  [Learn how to use a "Table of content" macro ](https://extensions.xwiki.org/xwiki/bin/view/Extension/TOC+Macro#HExample1:Simple)
* Discover how to create an application --> http://localhost:8080/xwiki/bin/view/Help/Applications/
*  [Learn how to use your application ](https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/FAQTutorial/FAQTutorialAWM#HCreatingnewFAQentries)
* [Discover how to add/remove extensions using the Extension Manager ](https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/CreatingExtensions/#HInstallinganExtension)
* [Discover the XWiki Administration UI ](https://extensions.xwiki.org/xwiki/bin/view/Extension/Administration%20Application#HDescription)

vmassol commented:
I like it! I think it brings value and not just for impaired viewers but for everyone.
Thanks

vmassol commented:
> Discover how to create an application --> http://localhost:8080/xwiki/bin/view/Help/Applications/

I think this should be a link on xwiki.org since we also need a link on https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/

BTW shouldn't we have the same vides on https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/ than on your screenshot? I see only 4 at https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/ which is weird.

vmassol commented:
For creating an app there's https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/CreatingABasicApp

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -42,50 +42,92 @@
-    $services.localization.render('help.videos.videoCard1.topic1'),
-    $services.localization.render('help.videos.videoCard1.topic2')
+    {
+      'url':'https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/',
+      'label':"$services.localization.render('help.videos.videoCard1.topic1')"
```
```suggestion
      'url':' https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/',
      'label': "$services.localization.render('help.videos.videoCard1.topic1')"
```

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -42,50 +42,92 @@
+      'label': "$services.localization.render('help.videos.videoCard5.topic1')"
+    },
+    {
+      'url':'https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/FAQTutorial/FAQTutorialAWM#HCreatingnewFAQentries',
+      'label':"$services.localization.render('help.videos.videoCard5.topic2')"
```
```suggestion
      'url': 'https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/FAQTutorial/FAQTutorialAWM#HCreatingnewFAQentries',
      'label': "$services.localization.render('help.videos.videoCard5.topic2')"
```

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -42,50 +42,92 @@
-    $services.localization.render('help.videos.videoCard6.topic1'),
-    $services.localization.render('help.videos.videoCard6.topic2')
+    {
+      'url':'https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/CreatingExtensions/#HInstallinganExtension',
+      'label':"$services.localization.render('help.videos.videoCard6.topic1')"
```
```suggestion
      'url': 'https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/CreatingExtensions/#HInstallinganExtension',
      'label': "$services.localization.render('help.videos.videoCard6.topic1')"
```

Sereza7 replied:
Addressed in https://github.com/xwiki/xwiki-platform/pull/2466/commits/ff76bbeae9f5b0f5eeb4dec94bd18c93e3e76cfe :+1:

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -42,50 +42,92 @@
+      'label':"$services.localization.render('help.videos.videoCard6.topic1')"
+    },
+    {
+      'url':'https://extensions.xwiki.org/xwiki/bin/view/Extension/Administration%20Application#HDescription',
+      'label':"$services.localization.render('help.videos.videoCard6.topic2')"
```
```suggestion
      'url': 'https://extensions.xwiki.org/xwiki/bin/view/Extension/Administration%20Application#HDescription',
      'label': "$services.localization.render('help.videos.videoCard6.topic2')"
```

Sereza7 replied:
Addressed in https://github.com/xwiki/xwiki-platform/pull/2466/commits/ff76bbeae9f5b0f5eeb4dec94bd18c93e3e76cfe :+1:

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -98,9 +140,16 @@
-          &lt;li&gt;$topic&lt;/li&gt;
+          &lt;li&gt;
+            &lt;a href="$topic.url"&gt;
+              $topic.label
```
I think there should be HTML escaping here.

Sereza7 replied:
Addressed in https://github.com/xwiki/xwiki-platform/pull/2466/commits/5a5e8ebed90bd22a5888290046ec60f69f664275 :+1:

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Code/Translations.xml:
```diff
@@ -192,6 +192,7 @@ help.videos.videoCard5.topic2=Learn how to use your application
 help.videos.videoCard6.title=Add / Remove Extensions
 help.videos.videoCard6.topic1=Discover how to add/remove extensions using the Extension Manager
 help.videos.videoCard6.topic2=Discover the XWiki Administration UI
+help.videos.hint=The video in this card is an alternative to the documentation behind those links:
```
I'm not sure a screen reader user is aware what "this card" is. Maybe rather something like:

```suggestion
help.videos.hint=The following documentation links provide an alternative to the preceeding video:
```

- but even then I'm not sure that it is clear to the user that there was a video, in particular as the iframe is marked with `role="presentation"`. What about putting this information in the title of the iframe and maybe even removing the `role="presentation"`?

Sereza7 replied:
Addressed in https://github.com/xwiki/xwiki-platform/pull/2466/commits/016b9d71aee7a2b9ef98122380e644bbe2aa399c :+1: 

I decided to remove the presentation role, and add the hint as an 'aria-description' to the iframe. This way, it can be read out to AT users and it's clearly concerning the content of this iframe. Updated the L10N value accordingly.

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -42,50 +42,92 @@
-    $services.localization.render('help.videos.videoCard5.topic1'),
-    $services.localization.render('help.videos.videoCard5.topic2')
+    {
+      'url': $xwiki.getURL('Help.Applications.WebHome'),
+      'label': "$services.localization.render('help.videos.videoCard5.topic1')"
```
Why is the `$xwiki.getURL` not in quotes but `$services.localization.render` is?

michitux reviewed (changes requested):
I have some doubts regarding the note. Also, the code style is quite inconsistent and I think there is missing escaping, see the detailed comments.

Sereza7 commented:
@vmassol 
> BTW shouldn't we have the same vides on https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/ than on your screenshot? I see only 4 at https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/ which is weird.

From what I could see, this page is custom content, none of the videos are the same.
This page uses a [video card macro](https://www.xwiki.org/xwiki/bin/view/Macros/VideoCard), which is very similar to the one I changed in this file. I updated the PR in https://github.com/xwiki/xwiki-platform/pull/2466/commits/291a7a1cc0c60db403f3775a9064f94423e5777b to ensure there's retrocompatibility of the velocity macro (even if not useful in this case).



> For creating an app there's https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/CreatingABasicApp
> 
> There's also https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/Creating%20A%20Contact%20Manager%20Application/

:+1: I chose to use the first one:
* We already used other steps of the getting started in other links here
* The video highlights app creation with AWM, so the first link is more appropriate

Updated in https://github.com/xwiki/xwiki-platform/pull/2466/commits/29f88791b7014717bee8931914c33070b9db8f4f

_____

All comments so far have been addressed.

vmassol commented:
> From what I could see, this page is custom content, none of the videos are the same.

hmm we might need to review this. I don't see how we would have enough manpower to have 2 sets of videos. We also need to check how much outdated both sets are.

> I chose to use the first one:

I think the second one has more value but I've now linked the first one to the second so I believe it's now ok, see https://www.xwiki.org/xwiki/bin/view/Documentation/UserGuide/GettingStarted/CreatingABasicApp?viewer=changes&rev1=7.2&rev2=8.1&

Thanks

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Code/Translations.xml:
```diff
@@ -192,6 +192,7 @@ help.videos.videoCard5.topic2=Learn how to use your application
 help.videos.videoCard6.title=Add / Remove Extensions
 help.videos.videoCard6.topic1=Discover how to add/remove extensions using the Extension Manager
 help.videos.videoCard6.topic2=Discover the XWiki Administration UI
+help.videos.hint=The following documentation links provide an alternative to this video.
```
Maybe we should make this an actual description of the video? Like saying that you see a screencast showcasing the features that are listed below without audio (which could be interesting information for a blind user)?

Sereza7 replied:
Addressed in https://github.com/xwiki/xwiki-platform/pull/2466/commits/758d90df8de85d58a87bea96d3438f98e33f458d :+1:

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -42,65 +42,114 @@
         #foreach ($topic in $data.topics)
-          &lt;li&gt;$topic&lt;/li&gt;
+          &lt;li&gt;#if ($topic.url)
+            &lt;a href="$escapetool.html($topic.url)"&gt;$topic.label&lt;/a&gt;
```
This needs to be `$escapetool.xml` as `$escapetool.html` doesn't escape `{`. Same for all other uses of `$escapetool.html`. Also, why is the label not escaped?

Sereza7 replied:
> Also, why is the label not escaped?

I went too fast when updating the definition of the topics. I thought I'd factorize the escaping but forgot to add it back.

:+1: Addressed in https://github.com/xwiki/xwiki-platform/pull/2466/commits/61140f947991b0dd48234aefdd51ba1254bf42ec

michitux commented on xwiki-platform-core/xwiki-platform-help/xwiki-platform-help-ui/src/main/resources/Help/Videos/WebHome.xml:
```diff
@@ -42,65 +42,114 @@
+          &lt;li&gt;#if ($topic.url)
+            &lt;a href="$escapetool.html($topic.url)"&gt;$topic.label&lt;/a&gt;
+          #else 
+            $topic
```
Same here, the topic should be escaped.

Sereza7 replied:
:+1: Addressed in https://github.com/xwiki/xwiki-platform/pull/2466/commits/61140f947991b0dd48234aefdd51ba1254bf42ec

Sereza7 commented:
Addressed all comments :)
